### PR TITLE
TUR-21619: Integrated CustomGroup into DSession event handling to avoid multiple node teardown/setyup between marks.

### PR DESCRIPTION
**Pull request by Heroico**
Added logging into scheduling internals. Changed Dsession logic for `CustomGroup` to reschedule strictly when a batch of tests is finished.

### CODE REVIEW

**Comment by Heroico** [comment]
There's a chance this introduces a new bug - still under investigation.

**Comment by Heroico** [comment]
The problem was in the upstream branch, only it was easier to reproduce with code in this branch

File: src/xdist/dsession.py [hunk 1]
```diff
@@ -205,6 +179,64 @@ def loop_once(self) -> None:
...
+    def is_node_finishing(self, node: WorkerController):
+        """Check if a test worker is considered to be finishing.
+
+        Evaluate whether it's on its last test, or if no tests are pending.
+        """
+        pending = self.sched.node2pending.get(node)
+        return pending is not None and len(pending) < 2
+
+
+    def is_node_clear(self, node: WorkerController):
```
Comment (ciaconet) [inline src/xdist/dsession.py:192]: This looks unused

File: src/xdist/dsession.py [hunk 2]
```diff
@@ -344,11 +394,15 @@ def worker_collectionfinish(
...
-                    self.terminal.write_line(
-                        f"scheduling tests via {self.sched.__class__.__name__}"
-                    )
-            self.actually_started = True
-            self.sched.schedule()
+                    self.report_line(f"[-] [dse] scheduling tests via {self.sched.__class__.__name__}")
+            if isinstance(self.sched, CustomGroup) and self.ready_to_run_tests and self.are_all_active_nodes_collected():
+                # we're coming back here after finishing a batch of tests - so start the next batch
+                self.reschedule()
+                self.reset_nodes_if_needed()
```
Comment (ciaconet) [inline src/xdist/dsession.py:408]: was this function definition removed?

File: src/xdist/dsession.py (hunk 1 above)
Comment (Heroico) [reply src/xdist/dsession.py:192]: It is unused, I mostly used it while debugging. I can remove it.

File: src/xdist/dsession.py [hunk 3]
```diff
@@ -205,6 +179,69 @@ def loop_once(self) -> None:
...
+
+    def are_all_active_nodes_collected(self):
+        """Check if all nodes have reported collection to be complete."""
+        if not all(n.gateway.id in self.worker_status for n in self._active_nodes):
+            return False
+        return all(self.worker_status[n.gateway.id] == "collected" for n in self._active_nodes)
+
+
+    def reset_nodes_if_needed(self):
+        if self.are_all_nodes_finishing() and self.ready_to_run_tests and not self.sched.do_resched:
```
Comment (ciaconet) [inline src/xdist/dsession.py:216]: With these changes it looks like `sched.do_resched` is no longer modified anywhere, and is by default `False`, should we just remove it?

Comment (Heroico) [reply src/xdist/dsession.py:216]: It should be reinstated, I removed it by mistake when squashing commits because I fail at everything

Comment (Heroico) [reply src/xdist/dsession.py:216]: They are reinstated now.

**Review by ciaconet — APPROVED** [review]

**Pull request merged**